<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Getting started · GhostKnockoffGWAS</title><meta name="title" content="Getting started · GhostKnockoffGWAS"/><meta property="og:title" content="Getting started · GhostKnockoffGWAS"/><meta property="twitter:title" content="Getting started · GhostKnockoffGWAS"/><meta name="description" content="Documentation for GhostKnockoffGWAS."/><meta property="og:description" content="Documentation for GhostKnockoffGWAS."/><meta property="twitter:description" content="Documentation for GhostKnockoffGWAS."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">GhostKnockoffGWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li class="is-active"><a class="tocitem" href>Getting started</a><ul class="internal"><li><a class="tocitem" href="#Typical-Workflow"><span>Typical Workflow</span></a></li><li><a class="tocitem" href="#Running-the-executable"><span>Running the executable</span></a></li><li><a class="tocitem" href="#Acceptable-Z-scores-format"><span>Acceptable Z-scores format</span></a></li></ul></li><li><a class="tocitem" href="../examples/">Examples</a></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li><li><a class="tocitem" href="../developer/">Developer documentation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Getting started</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Getting started</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS/blob/main/docs/src/man/getting_started.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Getting-started-with-Ghost-Knockoff-GWAS-analysis"><a class="docs-heading-anchor" href="#Getting-started-with-Ghost-Knockoff-GWAS-analysis">Getting started with Ghost Knockoff GWAS analysis</a><a id="Getting-started-with-Ghost-Knockoff-GWAS-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-started-with-Ghost-Knockoff-GWAS-analysis" title="Permalink"></a></h1><p>This tutorial is for generating <em>Ghost Knockoffs</em> for analyzing summary statistics from a genome-wide association studies (GWAS). The methodology is described in the following papers</p><blockquote><p>He Z, Liu L, Belloy ME, Le Guen Y, Sossin A, Liu X, Qi X, Ma S, Gyawali PK, Wyss-Coray T, Tang H. GhostKnockoff inference empowers identification of putative causal variants in genome-wide association studies. Nature Communications. 2022 Nov 23;13(1):7209.</p></blockquote><p>The main working assumption is that we do not have access to individual level genotype or phenotype data. Rather, for each SNP, we have</p><ol><li>Z-scores <span>$Z_j$</span> with respect to some phenotype from a GWAS, and</li><li>Access to LD (linkage disequilibrium) matrix</li></ol><h2 id="Typical-Workflow"><a class="docs-heading-anchor" href="#Typical-Workflow">Typical Workflow</a><a id="Typical-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Typical-Workflow" title="Permalink"></a></h2><p>Most users are expected to follow this workflow. For advanced users, see <a href="https://biona001.github.io/GhostKnockoffGWAS/dev/man/developer">Developer documentation</a>.</p><ol><li>Download the <a href>binary executable file</a> (XXX GB)</li><li>Download the <a href="https://drive.google.com/file/d/1_ajlxFWE2MCSgBXDgDbeZh9Lq721WANA/view">pre-computed knockoff statistics</a> (8.2GB)</li><li>Unzip both datasets</li><li>Prepare your input Z score file into accepted format, see <a href="https://biona001.github.io/GhostKnockoffGWAS/dev/man/getting_started/#Acceptable-Z-scores-format">Acceptable Z-scores</a> below. </li><li>Run the executable, see <a href="https://biona001.github.io/GhostKnockoffGWAS/dev/man/getting_started/#Running-the-executable">running the executable</a></li></ol><h2 id="Running-the-executable"><a class="docs-heading-anchor" href="#Running-the-executable">Running the executable</a><a id="Running-the-executable-1"></a><a class="docs-heading-anchor-permalink" href="#Running-the-executable" title="Permalink"></a></h2><p>To see required inputs (and optional inputs), invoke</p><pre><code class="language-shell hljs">$ ./GhostKnockoffGWAS --help

usage: &lt;PROGRAM&gt; [--genome-build GENOME-BUILD] [--m M] [--seed SEED]
                 [-h] zfile knockoff_dir Neffect out

positional arguments:
  zfile                 Tab or comma separated summary Z-score file.
                        The first row must be a header line that
                        contains CHR/POS/REF/ALT fields, as well as
                        something that indicates Z-score information.
                        CHR must be integer valued (e.g. chr21 is not
                        correct). If a column with `Z` is available,
                        it will be used as Z scores. Otherwise, we
                        need both `pvalue` and `beta` to be available,
                        or both `OR` (odds ratio) and `SE` (standard
                        error) to be available. In these cases, we
                        will convert them to Z scores.
  knockoff_dir          Path to the directory storing pre-processed
                        knockoff files
  Neffect               Effective sample size for original data (type:
                        Int64)
  out                   Output file prefix (without extensions)

optional arguments:
  --genome-build GENOME-BUILD
                        Specifies the human genome build for the
                        summary file. Must be 19 (hg19) or 38 (hg38).
                        (type: Int64, default: 38)
  --m M                 Number of knockoffs to generate (type: Int64,
                        default: 5)
  --seed SEED           Sets the random seed (type: Int64, default:
                        2023)
  -h, --help            show this help message and exit</code></pre><p>Example run:</p><pre><code class="language-shell hljs">./GhostKnockoffGWAS \
    ../../data/AD_Zscores_Meta_modified.txt \
    ../../data/EUR \
    506200 \
    ../../data/test_alzheimers_meta</code></pre><h2 id="Acceptable-Z-scores-format"><a class="docs-heading-anchor" href="#Acceptable-Z-scores-format">Acceptable Z-scores format</a><a id="Acceptable-Z-scores-format-1"></a><a class="docs-heading-anchor-permalink" href="#Acceptable-Z-scores-format" title="Permalink"></a></h2><p>The Z score file should satisfy the following requirements:</p><ol><li>It is a comma- or tab-separated file</li><li>The first row should be a header line</li><li>The header line should include <code>CHR</code>, <code>POS</code>, <code>REF</code>, <code>ALT</code>. The <code>ALT</code> allele will be treated as the effect allele and <code>REF</code> be treated as non-effect allele. The position of each variant must be from HG19 or HG38.</li><li>Z scores can be given in 3 ways<ol><li>If a column with the header <code>Z</code> exist, then it will be used as Z scores.</li><li>If <code>pvalue</code> and <code>beta</code> (effect size) columns both exist, we will transform them into Z scores</li><li>If <code>OR</code> (odds ratio) and <code>SE</code> (standard error) columns exist, they will be transformed into Z scores</li></ol></li></ol></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Home</a><a class="docs-footer-nextpage" href="../examples/">Examples »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Friday 22 December 2023 02:29">Friday 22 December 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
