
# Command-line documentation and usage of GhostKnockoffGWAS

## Usage

Simple run

```
GhostKnockoffGWAS --zfile example_zfile.txt --LD-files EUR --N 506200 --genome-build 38 --out example_output
```

## Required inputs

| Option name              | Argument        | Description   |
| :---                     |    :----:       |   :---        |
| `--zfile`        | String | Input file containing Z-scores as well as CHR/POS/REF/ALT. See [Acceptable Z-score files](https://biona001.github.io/GhostKnockoffGWAS/dev/man/zfile) for detailed requirement on this file. |
| `--LD-files`     | String | Input directory to the pre-processed LD files. Most users downloads this from the [Downloads Page](https://biona001.github.io/GhostKnockoffGWAS/dev/man/download) |
| `--N`            | Int    | Sample size for target (original) study |
| `--genome-build` | Int    | The human genome build used for SNP positions in `zfile` (this value must be 19 or 38) |
| `--out`          | String | Output file name (without extensions) |

## Optional inputs

| Option name              | Argument         | Description   |
| :---                    |    :----:         |   :---     |
| `--CHR`        | Int    | The column in `zfile` that will be read as chromosome number (note this must be an integer, e.g. chr22, X, chrX, ...etc are NOT acceptable). [If not specified, we will search for a column with header `CHR`] |
| `--POS`        | Int    | The column in `zfile` that will be read as SNP position . [If not specified, we will search for a column with header `POS`] |
| `--REF`        | Int    | The column in `zfile` that will be read as REF (non-effectiv) allele . [If not specified, we will search for a column with header `REF`] |
| `--ALT`        | Int    | The column in `zfile` that will be read as ALT (effective allele). [If not specified, we will search for a column with header `REF`] |
| `--Z`          | Int    | The column in `zfile` that will be read as Z-scores. [If not specified, we will search for a column with header `Z`] |
| `--seed`       | Int    | Sets the random seed [If not specified, defaults to `2023`] |
| `--verbose`    | Bool   | Whether to print intermediate messages [If not specified, defaults to `true`] |
| `--random-shuffle` | Bool | Whether to randomly permute the order of Z-scores and their knockoffs to adjust for potential ordering bias. [If not specified, defaults to `true`] |
| `--skip-shrinkage-check` | Bool | Whether to allow Knockoff analysis to proceed even with large (>0.25) LD shrinkages [If not specified, defaults to `false`] |


## Output format

1. A summary file, e.g. `example_output_summary.txt`. This file contains broad summary of the analysis
2. A comma-separated file, e.g. `example_output.txt`. This file contains the full `GhostKnockoffGWAS` output, one SNP in each row.
3. (optional) Manhattan plots, which can be generated by following [step 5 of detailed example](https://biona001.github.io/GhostKnockoffGWAS/dev/man/examples/#Step-5:-Generating-Manhattan-plots).

For a more detailed explanation on these 2 files, see [Tutorial](https://biona001.github.io/GhostKnockoffGWAS/dev/man/examples/#Step-4:-Interpreting-the-result). 
