{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command-line documentation and usage of GhostKnockoffGWAS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Usage\n",
    "\n",
    "Simple run\n",
    "\n",
    "```\n",
    "GhostKnockoffGWAS --zfile example_zfile.txt --LD-files EUR --N 506200 --genome-build 38 --out example_output\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Required inputs\n",
    "\n",
    "| Option name              | Argument        | Description   |\n",
    "| :---                     |    :----:       |   :---        |\n",
    "| `--zfile`        | String | Input file containing Z-scores as well as CHR/POS/REF/ALT. See [Acceptable Z-score files](https://biona001.github.io/GhostKnockoffGWAS/dev/man/zfile) for detailed requirement on this file. |\n",
    "| `--LD-files`     | String | Input directory to the pre-processed LD files. Most users downloads this from the [Downloads Page](https://biona001.github.io/GhostKnockoffGWAS/dev/man/download) |\n",
    "| `--N`            | Int    | Sample size for target (original) study |\n",
    "| `--genome-build` | Int    | The human genome build used for SNP positions in `zfile` (this value must be 19 or 38) |\n",
    "| `--out`          | String | Output file name (without extensions) |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optional inputs\n",
    "\n",
    "| Option name              | Argument         | Description   |\n",
    "| :---                    |    :----:         |   :---     |\n",
    "| `--CHR`        | Int    | The column in `zfile` that will be read as chromosome number (note this must be an integer, e.g. chr22, X, chrX, ...etc are NOT acceptable). [If not specified, we will search for a column with header `CHR`] |\n",
    "| `--POS`        | Int    | The column in `zfile` that will be read as SNP position . [If not specified, we will search for a column with header `POS`] |\n",
    "| `--REF`        | Int    | The column in `zfile` that will be read as REF (non-effectiv) allele . [If not specified, we will search for a column with header `REF`] |\n",
    "| `--ALT`        | Int    | The column in `zfile` that will be read as ALT (effective allele). [If not specified, we will search for a column with header `REF`] |\n",
    "| `--Z`          | Int    | The column in `zfile` that will be read as Z-scores. [If not specified, we will search for a column with header `Z`] |\n",
    "| `--seed`       | Int    | Sets the random seed [If not specified, defaults to `2023`] |\n",
    "| `--verbose`    | Bool   | Whether to print intermediate messages [If not specified, defaults to `true`] |\n",
    "| `--random-shuffle` | Bool | Whether to randomly permute the order of Z-scores and their knockoffs to adjust for potential ordering bias. [If not specified, defaults to `true`] |\n",
    "| `--skip-shrinkage-check` | Bool | Whether to allow Knockoff analysis to proceed even with large (>0.25) LD shrinkages [If not specified, defaults to `false`] |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output format\n",
    "\n",
    "1. A summary file, e.g. `example_output_summary.txt`. This file contains broad summary of the analysis\n",
    "2. A comma-separated file, e.g. `example_output.txt`. This file contains the full `GhostKnockoffGWAS` output, one SNP in each row.\n",
    "3. (optional) Manhattan plots, which can be generated by following [step 5 of detailed example](https://biona001.github.io/GhostKnockoffGWAS/dev/man/examples/#Step-5:-Generating-Manhattan-plots).\n",
    "\n",
    "For a more detailed explanation on these 2 files, see [Tutorial](https://biona001.github.io/GhostKnockoffGWAS/dev/man/examples/#Step-4:-Interpreting-the-result). "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.9.1",
   "language": "julia",
   "name": "julia-1.9"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
