<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Usage within Julia · GhostKnockoffGWAS</title><meta name="title" content="Usage within Julia · GhostKnockoffGWAS"/><meta property="og:title" content="Usage within Julia · GhostKnockoffGWAS"/><meta property="twitter:title" content="Usage within Julia · GhostKnockoffGWAS"/><meta name="description" content="Documentation for GhostKnockoffGWAS."/><meta property="og:description" content="Documentation for GhostKnockoffGWAS."/><meta property="twitter:description" content="Documentation for GhostKnockoffGWAS."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">GhostKnockoffGWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../getting_started/">Getting started</a></li><li><a class="tocitem" href="../examples/">Examples</a></li><li class="is-active"><a class="tocitem" href>Usage within Julia</a><ul class="internal"><li><a class="tocitem" href="#Usage-example"><span>Usage example</span></a></li><li><a class="tocitem" href="#Function-API"><span>Function API</span></a></li></ul></li><li><a class="tocitem" href="../video/">Video tutorials</a></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li><li><a class="tocitem" href="../developer/">Developer documentation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Usage within Julia</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Usage within Julia</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS/blob/main/docs/src/man/julia.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Usage-within-Julia"><a class="docs-heading-anchor" href="#Usage-within-Julia">Usage within Julia</a><a id="Usage-within-Julia-1"></a><a class="docs-heading-anchor-permalink" href="#Usage-within-Julia" title="Permalink"></a></h1><p><code>GhostKnockoffGWAS</code> is a regular Julia package, which can be used directly within Julia for greater flexibility. To install it, execute the following in Julia</p><pre><code class="language-julia hljs">using Pkg
Pkg.add(url=&quot;https://github.com/biona001/ghostbasil_jll.jl&quot;)
Pkg.add(url=&quot;https://github.com/biona001/Ghostbasil.jl&quot;)
Pkg.add(url=&quot;https://github.com/biona001/GhostKnockoffGWAS&quot;)</code></pre><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This package currently only works on Julia 1.8.x and 1.9.x. If you need it to work on a different Julia version, let us know by filing an issue on Github. </p></div></div><h2 id="Usage-example"><a class="docs-heading-anchor" href="#Usage-example">Usage example</a><a id="Usage-example-1"></a><a class="docs-heading-anchor-permalink" href="#Usage-example" title="Permalink"></a></h2><p>The following example performs summary-statistics GWAS under the GhostKnockoff framework.</p><pre><code class="language-julia hljs">using GhostKnockoffGWAS

# file paths and directories
knockoff_dir = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data/EUR&quot;
zfile = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data/AD_Zscores_Meta_modified.txt&quot;
outdir = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data&quot;
outfile = &quot;test_alzheimers_meta&quot;

# specify sample size and human genome build
N = 506200
hg_build = 38

# read Z-scores using built-in function read_zscores
z, chr, pos, effect_allele, non_effect_allele = GhostKnockoffGWAS.read_zscores(zfile)

# run analysis
@time ghostbasil_parallel(knockoff_dir, z, chr, pos, effect_allele, 
    non_effect_allele, N, hg_build, outdir, outname=outfile, 
    seed=2023)</code></pre><h2 id="Function-API"><a class="docs-heading-anchor" href="#Function-API">Function API</a><a id="Function-API-1"></a><a class="docs-heading-anchor-permalink" href="#Function-API" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.MK_statistics-Union{Tuple{T}, Tuple{Vector{T}, Array{Vector{T}, 1}}} where T" href="#GhostKnockoffGWAS.MK_statistics-Union{Tuple{T}, Tuple{Vector{T}, Array{Vector{T}, 1}}} where T"><code>GhostKnockoffGWAS.MK_statistics</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">MK_statistics(T0::Vector, Tk::Vector{Vector}; filter_method)</code></pre><p>Computes the multiple knockoff statistics kappa, tau, and W. </p><p><strong>Inputs</strong></p><ul><li><code>T0</code>: p-vector of importance score for original variables</li><li><code>Tk</code>: Vector storing T1, ..., Tm, where Ti is importance scores for    the <code>i</code>th knockoff copy</li><li><code>filter_method</code>: Either <code>Statistics.median</code> (default) or max (original    function used in 2019 Gimenez and Zou)</li></ul><p><strong>output</strong></p><ul><li><code>κ</code>: Index of the most significant feature (<code>κ[i] = 0</code> if original feature most    important, otherwise <code>κ[i] = k</code> if the <code>k</code>th knockoff is most important)</li><li><code>τ</code>: <code>τ[i]</code> stores the most significant statistic among original and knockoff   variables minus <code>filter_method()</code> applied to the remaining statistics. </li><li><code>W</code>: coefficient difference statistic <code>W[i] = abs(T0[i]) - abs(Tk[i])</code></li></ul><p><strong>Source</strong></p><p>https://github.com/biona001/Knockoffs.jl/blob/master/src/threshold.jl#L96</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/336f6ce6fe6669dcdf6211c05e913bab383b4929/src/utilities.jl#L275-L296">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.find_matching_indices-Tuple{AbstractVector, AbstractVector}" href="#GhostKnockoffGWAS.find_matching_indices-Tuple{AbstractVector, AbstractVector}"><code>GhostKnockoffGWAS.find_matching_indices</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">find_matching_indices(a, b)</code></pre><p>Returns output vector <code>c</code> such that <code>c[i]</code> contains all indices of <code>b</code> that  matches <code>a[i]</code>. Generalizes <code>indexin</code> in Base. </p><p><strong>Example</strong></p><pre><code class="language-julia hljs">a = [1, 2, 3, 2, 6]
b = [2, 1, 2, 4, 3]
c = find_matching_indices(a, b)
println(c)  # Output: [[2], [1, 3], [5], [1, 3], Int[]]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/336f6ce6fe6669dcdf6211c05e913bab383b4929/src/utilities.jl#L13-L26">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.ghost_knockoffs-Union{Tuple{T}, Tuple{AbstractVector{T}, AbstractMatrix{T}, AbstractMatrix{T}}} where T" href="#GhostKnockoffGWAS.ghost_knockoffs-Union{Tuple{T}, Tuple{AbstractVector{T}, AbstractMatrix{T}, AbstractMatrix{T}}} where T"><code>GhostKnockoffGWAS.ghost_knockoffs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ghost_knockoffs(Zscores, D, Σinv; [m=1])</code></pre><p>Generate Ghost knockoffs given a list of z-scores (GWAS summary statistic). </p><p><strong>Inputs</strong></p><ul><li><code>Zscores</code>: List of z-score statistics</li><li><code>D</code>: Matrix obtained from solving the knockoff problem satisfying    <code>(m+1)/m*Σ - D ⪰ 0</code></li><li><code>Σinv</code>: Inverse of the covariance matrix</li></ul><p><strong>optional inputs</strong></p><ul><li><code>m</code>: Number of knockoffs</li></ul><p><strong>Reference</strong></p><p>He, Z., Liu, L., Belloy, M. E., Le Guen, Y., Sossin, A., Liu, X., ... &amp; Ionita-Laza, I. (2021).  Summary statistics knockoff inference empowers identification of putative causal variants in  genome-wide association studies. </p><p><strong>Source</strong></p><p>https://github.com/biona001/Knockoffs.jl/blob/master/src/ghost.jl#L32</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/336f6ce6fe6669dcdf6211c05e913bab383b4929/src/utilities.jl#L241-L262">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.mk_threshold-Union{Tuple{T}, Tuple{Vector{T}, Vector{Int64}, Int64, Number}, Tuple{Vector{T}, Vector{Int64}, Int64, Number, Any}, Tuple{Vector{T}, Vector{Int64}, Int64, Number, Any, Int64}} where T&lt;:AbstractFloat" href="#GhostKnockoffGWAS.mk_threshold-Union{Tuple{T}, Tuple{Vector{T}, Vector{Int64}, Int64, Number}, Tuple{Vector{T}, Vector{Int64}, Int64, Number, Any}, Tuple{Vector{T}, Vector{Int64}, Int64, Number, Any, Int64}} where T&lt;:AbstractFloat"><code>GhostKnockoffGWAS.mk_threshold</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">mk_threshold(τ::Vector{T}, κ::Vector{Int}, m::Int, q::Number)</code></pre><p>Chooses the multiple knockoff threshold <code>τ̂ &gt; 0</code> by setting τ̂ = min{ t &gt; 0 : (1/m + 1/m * {#j: κ[j] ≥ 1 and W[j] ≥ t}) / {#j: κ[j] == 0 and W[j] ≥ τ̂} ≤ q }.</p><p><strong>Inputs</strong></p><ul><li><code>τ</code>: τ[i] stores the feature importance score for the ith feature, i.e. the value   T0 - median(T1,...,Tm). Note in Gimenez and Zou, the max function is used    instead of median</li><li><code>κ</code>: κ[i] stores which of m knockoffs has largest importance score. When original    variable has largest score, κ[i] == 0.</li><li><code>m</code>: Number of knockoffs per variable generated</li><li><code>q</code>: target FDR (between 0 and 1)</li><li><code>rej_bounds</code>: Number of values of top τ to consider (default = 10000)</li></ul><p><strong>Reference:</strong></p><ul><li>Equations 8 and 9 in supplement of &quot;Identification of putative causal loci in    wholegenome sequencing data via knockoff statistics&quot; by He et al. </li><li>Algorithm 1 of &quot;Improving the Stability of the Knockoff Procedure: Multiple    Simultaneous Knockoffs and Entropy Maximization&quot; by Gimenez and Zou.</li></ul><p><strong>Source</strong></p><p>https://github.com/biona001/Knockoffs.jl/blob/master/src/threshold.jl#L55</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/336f6ce6fe6669dcdf6211c05e913bab383b4929/src/utilities.jl#L323-L347">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.read_zscores-Tuple{String}" href="#GhostKnockoffGWAS.read_zscores-Tuple{String}"><code>GhostKnockoffGWAS.read_zscores</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">read_zscores(filepath::String)</code></pre><p>Helper function to read a Z-score file. First row must be a header column with  CHR, POS, REF, ALT, and Z. All other columns will be ignored. </p><p><strong>todo: detect duplicate SNPs</strong></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/336f6ce6fe6669dcdf6211c05e913bab383b4929/src/utilities.jl#L127-L134">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.sample_mvn_efficient-Union{Tuple{T}, Tuple{AbstractMatrix{T}, AbstractMatrix{T}, Int64}} where T" href="#GhostKnockoffGWAS.sample_mvn_efficient-Union{Tuple{T}, Tuple{AbstractMatrix{T}, AbstractMatrix{T}, Int64}} where T"><code>GhostKnockoffGWAS.sample_mvn_efficient</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sample_mvn_efficient(C::AbstractMatrix{T}, D::AbstractMatrix{T}, m::Int)</code></pre><p>Efficiently samples from <code>N(0, A)</code> where</p><p class="math-container">\[\begin{aligned}
A &amp;= \begin{pmatrix}
    C &amp; C-D &amp; \cdots &amp; C-D\\
    C-D &amp; C &amp; \cdots &amp; C-D\\
    \vdots &amp; &amp; \ddots &amp; \vdots\\
    C-D &amp; C-D &amp; &amp; C
\end{pmatrix}
\end{aligned}\]</p><p>Note there are <code>m</code> blocks per row/col</p><p><strong>Source</strong></p><p>https://github.com/biona001/Knockoffs.jl/blob/master/src/ghost.jl#L60</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/336f6ce6fe6669dcdf6211c05e913bab383b4929/src/utilities.jl#L205-L223">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../examples/">« Examples</a><a class="docs-footer-nextpage" href="../video/">Video tutorials »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Tuesday 16 January 2024 20:04">Tuesday 16 January 2024</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
