<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Usage within Julia · GhostKnockoffGWAS</title><meta name="title" content="Usage within Julia · GhostKnockoffGWAS"/><meta property="og:title" content="Usage within Julia · GhostKnockoffGWAS"/><meta property="twitter:title" content="Usage within Julia · GhostKnockoffGWAS"/><meta name="description" content="Documentation for GhostKnockoffGWAS."/><meta property="og:description" content="Documentation for GhostKnockoffGWAS."/><meta property="twitter:description" content="Documentation for GhostKnockoffGWAS."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">GhostKnockoffGWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../intro/">Introduction</a></li><li><a class="tocitem" href="../documentation/">Documentation</a></li><li><a class="tocitem" href="../download/">Downloads</a></li><li><a class="tocitem" href="../zfile/">Acceptable Z-scores file</a></li><li><a class="tocitem" href="../examples/">Tutorial</a></li><li><a class="tocitem" href="../solveblocks/">Customizing LD files</a></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li><li class="is-active"><a class="tocitem" href>Usage within Julia</a><ul class="internal"><li><a class="tocitem" href="#Usage-example"><span>Usage example</span></a></li><li><a class="tocitem" href="#Function-API"><span>Function API</span></a></li></ul></li><li><a class="tocitem" href="../developer/">Developer documentation</a></li><li><a class="tocitem" href="../gallery/">Gallery</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Usage within Julia</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Usage within Julia</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS/blob/main/docs/src/man/julia.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Usage-within-Julia"><a class="docs-heading-anchor" href="#Usage-within-Julia">Usage within Julia</a><a id="Usage-within-Julia-1"></a><a class="docs-heading-anchor-permalink" href="#Usage-within-Julia" title="Permalink"></a></h1><p><code>GhostKnockoffGWAS</code> is a regular Julia package, which can be used directly within Julia for greater flexibility. To install it, execute the following in Julia</p><pre><code class="language-julia hljs">using Pkg
Pkg.add(url=&quot;https://github.com/biona001/ghostbasil_jll.jl&quot;)
Pkg.add(url=&quot;https://github.com/biona001/Ghostbasil.jl&quot;)
Pkg.add(url=&quot;https://github.com/biona001/HDF5.jl&quot;) # needed to resolve https://github.com/biona001/GhostKnockoffGWAS/issues/7
Pkg.add(url=&quot;https://github.com/biona001/GhostKnockoffGWAS&quot;)</code></pre><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This package currently only works on Linux machines with Julia 1.8.x, 1.9.x, and 1.10.0. If you need it to work on a different Julia version, let us know by filing an issue on Github. </p></div></div><h2 id="Usage-example"><a class="docs-heading-anchor" href="#Usage-example">Usage example</a><a id="Usage-example-1"></a><a class="docs-heading-anchor-permalink" href="#Usage-example" title="Permalink"></a></h2><p>The following example performs summary-statistics GWAS under the GhostKnockoff framework.</p><pre><code class="language-julia hljs">using GhostKnockoffGWAS

# file paths and directories
LD_files = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data/EUR&quot;
zfile = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data/AD_Zscores_Meta_modified.txt&quot;
outdir = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data&quot;

# specify sample size and human genome build
N = 506200
hg_build = 38

# read Z-scores using built-in function read_zscores
z, chr, pos, effect_allele, non_effect_allele = GhostKnockoffGWAS.read_zscores(zfile)

# run analysis
@time ghostknockoffgwas(LD_files, z, chr, pos, effect_allele, 
    non_effect_allele, N, hg_build, outdir, outname=&quot;test_alzheimers_meta&quot;)</code></pre><h2 id="Function-API"><a class="docs-heading-anchor" href="#Function-API">Function API</a><a id="Function-API-1"></a><a class="docs-heading-anchor-permalink" href="#Function-API" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.ghostknockoffgwas" href="#GhostKnockoffGWAS.ghostknockoffgwas"><code>GhostKnockoffGWAS.ghostknockoffgwas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ghostknockoffgwas(LD_files::String, z::Vector{Float64}, chr::Vector{Int}, 
    effect_allele::Vector{String}, non_effect_allele::Vector{String}, N::Int,
    hg_build::Int, outdir::String; [outname=&quot;result&quot;], [seed=2023], 
    [target_chrs=sort!(unique(chr))], [A_scaling_factor = 0.01], 
    [kappa_lasso=0.6], [LD_shrinkage=false], 
    [target_fdrs=[0.01, 0.05, 0.1, 0.2]], [verbose=true], 
    [skip_shrinkage_check=false])</code></pre><p>Runs the main <code>GhostKnockoffGWAS</code> pipeline on the Z scores in <code>z</code> using  pre-computed knockoff data in <code>LD_files</code>. </p><p><strong>Inputs</strong></p><ul><li><code>LD_files</code>: Directory that stores pre-computed knockoff results</li><li><code>z</code>: Vector of Z scores</li><li><code>chr</code>: Chromosome of each Z score (cannot be X/Y/M chromosomes)</li><li><code>pos</code>: Position of each Z score (specify hg build with <code>hg_build</code>)</li><li><code>effect_allele</code>: Effect allele of Z score (i.e. ALT)</li><li><code>non_effect_allele</code>: Non-effect allele of Z score (i.e. REF)</li><li><code>N</code>: sample size of original study</li><li><code>hg_build</code>: Human genome build (must be 19 or 38)</li><li><code>outdir</code>: output directory, which must exist. We will output 2 files, one    containing the full analysis results, as well as a summary file. </li></ul><p><strong>Optional inputs</strong></p><ul><li><code>seed</code>: Random seed for reproducibility (default = <code>2023</code>)</li><li><code>target_chrs</code>: Target chromosomes to analyze. For example, one can specify   <code>target_chrs = 22</code> to only analyze 1 chromosome, or <code>target_chrs = [1, 2]</code>   to only analyze 2 chromosomes (default = <code>sort!(unique(chr))</code>).</li><li><code>A_scaling_factor</code>: The scaling factor for <code>A = [X X̃]&#39;*[X X̃]</code> for improving   numerical stability. Scaling proceeds by adding <code>A_scaling_factor*I</code> to <code>A</code>   (default = 0.01). </li><li><code>kappa_lasso</code>: A constant between 0 and 1 for tuning Lasso&#39;s lambda path.    Larger value forces earlier exit in Lasso lambda path, resulting in stronger    shrinkage. See the &quot;lasso-min&quot; method of &quot;Controlled Variable Selection from   Summary Statistics Only? A Solution via GhostKnockoffs and Penalized    Regression&quot; by Chen et al (default <code>0.6</code>).</li><li><code>LD_shrinkage</code>: Whether to perform shrinkage to LD and S matrices following   method in SuSiE paper (i.e. eq 24 of &quot;Fine-mapping from summary data with    the “Sum of Single Effects” model&quot; by Zou et al). If <code>false</code>, we will still   compute the shrinkage level, but it will not be used to adjust the LD   matrices (default <code>false</code>). </li><li><code>target_fdrs</code>: Default target FDR levels (default = [0.01, 0.05, 0.1, 0.2])</li><li><code>verbose</code>: Whether to print progress and informative intermediate results (   default = <code>true</code>)</li><li><code>skip_shrinkage_check</code>: Forces a result output even if there is a high   estimated LD shrinkage by SuSiE&#39;s method (default = <code>false</code>)</li><li><code>random_shuffle</code>: Whether to randomly permute the order of the original and    knockoff variables (default <code>false</code>). The main purpose of this option is    to take care of potential ordering bias of Lasso solvers. However, in our    simulations we never observed such biases, so we turn this off by default. </li></ul><p><strong>Output</strong></p><p>By default we output 2 files into <code>outdir</code></p><ul><li>A knockoff statistics file where each SNP occupies a row and the columns include    various SNP attributes include rsid, AF, chr, pos, zscores...etc. The    columns <code>selected_fdr_FDR</code> indicates whether the variant was ultimately   selected under the false discovery rate threshold of <code>FDR</code>.</li><li>A summary statistics file. The first dozens of rows print, for each false    discovery rate threshold <code>FDR</code>, the knockoff threshold <code>τ̂</code> and the number of   groups that pass this threshold. The next couple of lines print some    parameters used in the knockoff analysis, as well as some timing results. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/57835f7e1222c0f2996d47f00216efa47f5bcb44/src/ghostbasil_parallel.jl#L1-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.read_zscores" href="#GhostKnockoffGWAS.read_zscores"><code>GhostKnockoffGWAS.read_zscores</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">read_zscores(filepath::String)</code></pre><p>Helper function to read a Z-score file at <code>filepath</code>. This function is mainly  intended for Julia users running <code>GhostKnockoffGWAS</code> in the REPL.</p><p><strong>Input</strong></p><ul><li><code>filepath</code>: Full file path to the Z-score file. First row must be a header   column with <code>CHR</code>, <code>POS</code>, <code>REF</code>, <code>ALT</code>, and <code>Z</code>. If the file contains these    information but has a different header name for them, use the optional input   arguments below. All other columns will be ignored.</li></ul><p><strong>Optional inputs</strong></p><ul><li><code>chr_col</code>: An integer, specifying which column in <code>filepath</code> should be read as   CHR (by default we search for a header <code>CHR</code>)</li><li><code>pos_col</code>: An integer, specifying which column in <code>filepath</code> should be read as    POS (by default we search for a header <code>POS</code>)</li><li><code>ref_col</code>: An integer, specifying which column in <code>filepath</code> should be read as    REF (by default we search for a header <code>REF</code>)</li><li><code>alt_col</code>: An integer, specifying which column in <code>filepath</code> should be read as   ALT (by default we search for a header <code>ALT</code>)</li><li><code>z_col</code>: An integer, specifying which column in <code>filepath</code> should be read as Z   (by default we search for a header <code>Z</code>)</li></ul><p><strong>Output</strong></p><ul><li><code>z</code>: The Z scores stored in the <code>Z</code> column of <code>filepath</code></li><li><code>chr</code>: The chromosome number stored in <code>CHR</code> column of <code>filepath</code>. Only integer   values are allowed.</li><li><code>pos</code>: The position number stored in <code>POS</code> column of <code>filepath</code>.</li><li><code>effect_allele</code>: The allele stored in <code>ALT</code> column of <code>filepath</code>.</li><li><code>non_effect_allele</code>: The allele stored in <code>REF</code> column of <code>filepath</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/57835f7e1222c0f2996d47f00216efa47f5bcb44/src/utilities.jl#L142-L173">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.solve_blocks" href="#GhostKnockoffGWAS.solve_blocks"><code>GhostKnockoffGWAS.solve_blocks</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">solve_blocks(vcffile::String, chr::Int, start_bp::Int, end_bp::Int, 
    outdir::String, hg_build::Int; [m=5], [tol=0.0001], [min_maf=0.01], 
    [min_hwe=0.0], [force_block_diag=true], 
    [method::String = &quot;maxent&quot;], [linkage::String=&quot;average&quot;],
    [force_contiguous::Bool=false], [group_cor_cutoff::Float64=0.5], 
    [group_rep_cutoff::Float64=0.5], [verbose=true])</code></pre><p>Solves the group knockoff optimization problem on provided individual-level data and outputs the result into <code>outdir</code>. All variants that reside on chromosome  <code>chr</code> with position between <code>start_bp</code> and <code>end_bp</code> (inclusive) will be included. </p><p><strong>Note on large VCF files</strong></p><p>Currently reading/parsing a VCF file is a single-threaded operation (even if  it is indexed). Thus, we <em>strongly recommend</em> one to split the input VCF  file by chromosomes, and possibly into smaller chunks, before running this function. </p><p><strong>Inputs</strong></p><ul><li><code>vcffile</code>: A VCF file storing individual level genotypes. Must end in <code>.vcf</code>    or <code>.vcf.gz</code>. The ALT field for each record must be unique, i.e.    multiallelic records must be split first. Missing genotypes will be imputed   by column mean. </li><li><code>chr</code>: Target chromosome. This MUST be an integer and it must match the <code>CHROM</code>   field in your VCF file. Thus, if your VCF file has CHROM field like <code>chr1</code>,    <code>CHR1</code>, or <code>CHROM1</code> etc, each record must be renamed into <code>1</code>. </li><li><code>start_bp</code>: starting basepair (position)</li><li><code>end_bp</code>: ending basepair (position)</li><li><code>outdir</code>: Directory that the output will be stored in (must exist)</li><li><code>hg_build</code>: human genome build for the VCF file, must be 19 (hg19) or 38 (hg38)</li></ul><p><strong>Optional inputs (for group knockoff optimization)</strong></p><ul><li><code>snps_to_keep</code>: Vector of SNP positions to import. If specified, only SNPs   whose position is listed in <code>snps_to_keep</code> will be kept (default <code>nothing</code>)</li><li><code>tol</code>: Convergence tolerlance for coordinate descent algorithm (default <code>0.0001</code>)</li><li><code>min_maf</code>: Minimum minor allele frequency for a variable to be considered (   default <code>0.01</code>)</li><li><code>min_hwe</code>: Cutoff for hardy-weinburg equilibrium p-values. Only SNPs with    p-value &gt; <code>min_hwe</code> will be included (default <code>0.0</code>)</li><li><code>force_block_diag</code>: Whether to re-order the columns/rows of the correlation    matrix and corresponding <code>S</code> matrix so that features in the same group    are contiguous (default <code>true</code>). This has no impact on the final results,    it is simply for computational performance. </li><li><code>method</code>: group knockoff optimization algorithm, choices include &quot;maxent&quot;    (default), &quot;mvr&quot;, &quot;sdp&quot;, or &quot;equi&quot;. See sec 2 of https://arxiv.org/abs/2310.15069</li><li><code>linkage</code>: <em>cluster linkage</em> function to use for hierarchically clustering groups.   It defines how the distances between features are aggregated into the distances    between groups. Valid choices include:<ul><li><code>:average</code> (default): use the mean distance between any of the cluster members</li><li><code>:single</code>: use the minimum distance between any of the cluster members</li><li><code>:complete</code>: use the maximum distance between any of the members</li><li><code>:ward</code>: the distance is the increase of the average squared distance of a   point to its cluster centroid after merging the two clusters</li><li><code>:ward_presquared</code>: same as <code>:ward</code>, but assumes that the distances in d    are already squared.</li></ul></li><li><code>force_contiguous</code>: whether to force groups to be contiguous (default <code>false</code>).   Note if <code>force_contiguous=true</code>, <code>linkage</code> must be <code>:single</code>)</li><li><code>group_cor_cutoff</code>: correlation cutoff value for defining groups (default    <code>0.5</code>). Value should be between 0 and 1, where larger values correspond to    larger groups. </li><li><code>group_rep_cutoff</code>: cutoff value for selecting group-representatives (default   <code>0.5</code>). Value should be between 0 and 1, where larger values correspond to    more representatives per group. </li><li><code>verbose</code>: whether to print informative intermediate results (default <code>true</code>)</li></ul><p><strong>output</strong></p><p>Calling <code>solve_blocks</code> will create 3 files in the directory <code>outdir/chr</code>:</p><ul><li><code>XXX.h5</code>:  This contains data (Sigma, S, groups, ..., etc) for region XXX. It    contains the following:<ul><li><code>D</code>: A <code>p × p</code> (dense) matrix corresponding to the S matrix for both the   representative and non-representative variables. Knockoff sampling should    use this matrix. </li><li><code>S</code>: Matrix obtained from solving the group knockoff optimization problem    on the representative (group-key) variables.</li><li><code>Sigma</code>: The original <code>p × p</code> correlation matrix estimated from <code>vcffile</code></li><li><code>SigmaInv</code>: Inverse of <code>Sigma</code></li><li><code>Sigma_reps</code>: The correlation matrix for the representative variables. This   is the matrix actually used to solve the group-knockoff optimization</li><li><code>Sigma_reps_inv</code>: Inverse of <code>Sigma_reps</code></li><li><code>group_reps</code>: Indices <code>groups</code> that are used as representatives (i.e.    group-key variables)</li><li><code>groups</code>: The group membership vector</li></ul></li><li><code>Info_XXX.csv</code>: This includes information for each variant (chr/pos/etc) present    in the corresponding <code>.h5</code> file.</li><li><code>summary_XXX.csv</code>: Summary file for the knockoff optimization problem</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/57835f7e1222c0f2996d47f00216efa47f5bcb44/src/make_hdf5.jl#L158-L243">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../FAQ/">« FAQ</a><a class="docs-footer-nextpage" href="../developer/">Developer documentation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Thursday 27 June 2024 22:12">Thursday 27 June 2024</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
