<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Usage within Julia · GhostKnockoffGWAS</title><meta name="title" content="Usage within Julia · GhostKnockoffGWAS"/><meta property="og:title" content="Usage within Julia · GhostKnockoffGWAS"/><meta property="twitter:title" content="Usage within Julia · GhostKnockoffGWAS"/><meta name="description" content="Documentation for GhostKnockoffGWAS."/><meta property="og:description" content="Documentation for GhostKnockoffGWAS."/><meta property="twitter:description" content="Documentation for GhostKnockoffGWAS."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">GhostKnockoffGWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../intro/">Introduction</a></li><li><a class="tocitem" href="../documentation/">Documentation</a></li><li><a class="tocitem" href="../download/">Downloads</a></li><li><a class="tocitem" href="../zfile/">Acceptable Z-scores file</a></li><li><a class="tocitem" href="../examples/">Tutorial</a></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li><li class="is-active"><a class="tocitem" href>Usage within Julia</a><ul class="internal"><li><a class="tocitem" href="#Usage-example"><span>Usage example</span></a></li><li><a class="tocitem" href="#Function-API"><span>Function API</span></a></li><li><a class="tocitem" href="#Compiling-GhostKnockoffGWAS"><span>Compiling GhostKnockoffGWAS</span></a></li></ul></li><li><a class="tocitem" href="../developer/">Developer documentation</a></li><li><a class="tocitem" href="../gallery/">Gallery</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Usage within Julia</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Usage within Julia</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS/blob/main/docs/src/man/julia.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Usage-within-Julia"><a class="docs-heading-anchor" href="#Usage-within-Julia">Usage within Julia</a><a id="Usage-within-Julia-1"></a><a class="docs-heading-anchor-permalink" href="#Usage-within-Julia" title="Permalink"></a></h1><p><code>GhostKnockoffGWAS</code> is a regular Julia package, which can be used directly within Julia for greater flexibility. To install it, execute the following in Julia</p><pre><code class="language-julia hljs">using Pkg
Pkg.add(url=&quot;https://github.com/biona001/ghostbasil_jll.jl&quot;)
Pkg.add(url=&quot;https://github.com/biona001/Ghostbasil.jl&quot;)
Pkg.add(url=&quot;https://github.com/biona001/HDF5.jl&quot;) # needed to resolve https://github.com/biona001/GhostKnockoffGWAS/issues/7
Pkg.add(url=&quot;https://github.com/biona001/GhostKnockoffGWAS&quot;)</code></pre><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>This package currently only works on Linux machines with Julia 1.8.x, 1.9.x, and 1.10.0. If you need it to work on a different Julia version, let us know by filing an issue on Github. </p></div></div><h2 id="Usage-example"><a class="docs-heading-anchor" href="#Usage-example">Usage example</a><a id="Usage-example-1"></a><a class="docs-heading-anchor-permalink" href="#Usage-example" title="Permalink"></a></h2><p>The following example performs summary-statistics GWAS under the GhostKnockoff framework.</p><pre><code class="language-julia hljs">using GhostKnockoffGWAS

# file paths and directories
LD_files = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data/EUR&quot;
zfile = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data/AD_Zscores_Meta_modified.txt&quot;
outdir = &quot;/home/groups/sabatti/.julia/dev/GhostKnockoffGWAS/data&quot;

# specify sample size and human genome build
N = 506200
hg_build = 38

# read Z-scores using built-in function read_zscores
z, chr, pos, effect_allele, non_effect_allele = GhostKnockoffGWAS.read_zscores(zfile)

# run analysis
@time ghostknockoffgwas(LD_files, z, chr, pos, effect_allele, 
    non_effect_allele, N, hg_build, outdir, outname=&quot;test_alzheimers_meta&quot;)</code></pre><h2 id="Function-API"><a class="docs-heading-anchor" href="#Function-API">Function API</a><a id="Function-API-1"></a><a class="docs-heading-anchor-permalink" href="#Function-API" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.ghostknockoffgwas" href="#GhostKnockoffGWAS.ghostknockoffgwas"><code>GhostKnockoffGWAS.ghostknockoffgwas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ghostknockoffgwas(LD_files::String, z::Vector{Float64}, chr::Vector{Int}, 
    effect_allele::Vector{String}, non_effect_allele::Vector{String}, N::Int,
    hg_build::Int, outdir::String; [outname=&quot;result&quot;], [seed=2023], 
    [target_chrs=sort!(unique(chr))], [A_scaling_factor = 0.01], 
    [kappa_lasso=0.6], [LD_shrinkage=false], 
    [target_fdrs=[0.01, 0.05, 0.1, 0.2]], [verbose=true], 
    [skip_shrinkage_check=false])</code></pre><p>Runs the main <code>GhostKnockoffGWAS</code> pipeline on the Z scores in <code>z</code> using  pre-computed knockoff data in <code>LD_files</code>. </p><p><strong>Inputs</strong></p><ul><li><code>LD_files</code>: Directory that stores pre-computed knockoff results</li><li><code>z</code>: Vector of Z scores</li><li><code>chr</code>: Chromosome of each Z score (cannot be X/Y/M chromosomes)</li><li><code>pos</code>: Position of each Z score (specify hg build with <code>hg_build</code>)</li><li><code>effect_allele</code>: Effect allele of Z score (i.e. ALT)</li><li><code>non_effect_allele</code>: Non-effect allele of Z score (i.e. REF)</li><li><code>N</code>: sample size of original study</li><li><code>hg_build</code>: Human genome build (must be 19 or 38)</li><li><code>outdir</code>: output directory, which must exist. We will output 2 files, one    containing the full analysis results, as well as a summary file. </li></ul><p><strong>Optional inputs</strong></p><ul><li><code>seed</code>: Random seed for reproducibility (default = <code>2023</code>)</li><li><code>target_chrs</code>: Target chromosomes to analyze. For example, one can specify   <code>target_chrs = 22</code> to only analyze 1 chromosome, or <code>target_chrs = [1, 2]</code>   to only analyze 2 chromosomes (default = <code>sort!(unique(chr))</code>).</li><li><code>A_scaling_factor</code>: The scaling factor for <code>A = [X X̃]&#39;*[X X̃]</code> for improving   numerical stability. Scaling proceeds by adding <code>A_scaling_factor*I</code> to <code>A</code>   (default = 0.01). </li><li><code>kappa_lasso</code>: A constant between 0 and 1 for tuning Lasso&#39;s lambda path.    Larger value forces earlier exit in Lasso lambda path, resulting in stronger    shrinkage. See the &quot;lasso-min&quot; method of &quot;Controlled Variable Selection from   Summary Statistics Only? A Solution via GhostKnockoffs and Penalized    Regression&quot; by Chen et al (default <code>0.6</code>).</li><li><code>LD_shrinkage</code>: Whether to perform shrinkage to LD and S matrices following   method in SuSiE paper (i.e. eq 24 of &quot;Fine-mapping from summary data with    the “Sum of Single Effects” model&quot; by Zou et al). If <code>false</code>, we will still   compute the shrinkage level, but it will not be used to adjust the LD   matrices (default <code>false</code>). </li><li><code>target_fdrs</code>: Default target FDR levels (default = [0.01, 0.05, 0.1, 0.2])</li><li><code>verbose</code>: Whether to print progress and informative intermediate results (   default = <code>true</code>)</li><li><code>skip_shrinkage_check</code>: Forces a result output even if there is a high   estimated LD shrinkage by SuSiE&#39;s method (default = <code>false</code>)</li></ul><p><strong>Output</strong></p><p>By default we output 2 files into <code>outdir</code></p><ul><li>A knockoff statistics file where each SNP occupies a row and the columns include    various SNP attributes include rsid, AF, chr, pos, zscores...etc. The    columns <code>selected_fdr_FDR</code> indicates whether the variant was ultimately   selected under the false discovery rate threshold of <code>FDR</code>.</li><li>A summary statistics file. The first dozens of rows print, for each false    discovery rate threshold <code>FDR</code>, the knockoff threshold <code>τ̂</code> and the number of   groups that pass this threshold. The next couple of lines print some    parameters used in the knockoff analysis, as well as some timing results. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/6e97b701869cb6ec338b5f373215c26381aa744f/src/ghostbasil_parallel.jl#L1-L59">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GhostKnockoffGWAS.read_zscores" href="#GhostKnockoffGWAS.read_zscores"><code>GhostKnockoffGWAS.read_zscores</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">read_zscores(filepath::String)</code></pre><p>Helper function to read a Z-score file at <code>filepath</code>. This function is mainly  intended for Julia users running <code>GhostKnockoffGWAS</code> in the REPL.</p><p><strong>Input</strong></p><ul><li><code>filepath</code>: Full file path to the Z-score file. First row must be a header   column with <code>CHR</code>, <code>POS</code>, <code>REF</code>, <code>ALT</code>, and <code>Z</code>. If the file contains these    information but has a different header name for them, use the optional input   arguments below. All other columns will be ignored.</li></ul><p><strong>Optional inputs</strong></p><ul><li><code>chr_col</code>: An integer, specifying which column in <code>filepath</code> should be read as   CHR (by default we search for a header <code>CHR</code>)</li><li><code>pos_col</code>: An integer, specifying which column in <code>filepath</code> should be read as    POS (by default we search for a header <code>POS</code>)</li><li><code>ref_col</code>: An integer, specifying which column in <code>filepath</code> should be read as    REF (by default we search for a header <code>REF</code>)</li><li><code>alt_col</code>: An integer, specifying which column in <code>filepath</code> should be read as   ALT (by default we search for a header <code>ALT</code>)</li><li><code>z_col</code>: An integer, specifying which column in <code>filepath</code> should be read as Z   (by default we search for a header <code>Z</code>)</li></ul><p><strong>Output</strong></p><ul><li><code>z</code>: The Z scores stored in the <code>Z</code> column of <code>filepath</code></li><li><code>chr</code>: The chromosome number stored in <code>CHR</code> column of <code>filepath</code>. Only integer   values are allowed.</li><li><code>pos</code>: The position number stored in <code>POS</code> column of <code>filepath</code>.</li><li><code>effect_allele</code>: The allele stored in <code>ALT</code> column of <code>filepath</code>.</li><li><code>non_effect_allele</code>: The allele stored in <code>REF</code> column of <code>filepath</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/GhostKnockoffGWAS/blob/6e97b701869cb6ec338b5f373215c26381aa744f/src/utilities.jl#L142-L173">source</a></section></article><h2 id="Compiling-GhostKnockoffGWAS"><a class="docs-heading-anchor" href="#Compiling-GhostKnockoffGWAS">Compiling GhostKnockoffGWAS</a><a id="Compiling-GhostKnockoffGWAS-1"></a><a class="docs-heading-anchor-permalink" href="#Compiling-GhostKnockoffGWAS" title="Permalink"></a></h2><p>I compiled this with julia v1.9.0 on Sherlock cluster, with <code>gcc/7.3.0</code> loaded. </p><ol><li>Within julia,  <code>]add libcxxwrap_julia_jll</code>  Note: as of Feb 2024, <code>libcxxwrap_julia_jll</code> must be v0.11.x</li><li>Make sure <code>GhostKnockoffGWAS</code> is installed within Julia. </li><li><code>dev</code> the package via  <code>julia  ]dev GhostKnockoffGWAS</code></li><li>Compile using <a href="https://github.com/JuliaLang/PackageCompiler.jl">PackageCompiler.jl</a></li></ol><pre><code class="language-julia hljs">using PackageCompiler, GhostKnockoffGWAS
src = normpath(pathof(GhostKnockoffGWAS), &quot;../..&quot;)
des = normpath(pathof(GhostKnockoffGWAS), &quot;../../app_linux_x86&quot;)
precompile_script = normpath(pathof(GhostKnockoffGWAS), &quot;../precompile.jl&quot;)
@time create_app(src, des, 
    include_lazy_artifacts=true, 
    force=true, 
    precompile_execution_file=precompile_script
)</code></pre><p>The last step takes &gt;15 minutes. </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../FAQ/">« FAQ</a><a class="docs-footer-nextpage" href="../developer/">Developer documentation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.0 on <span class="colophon-date" title="Friday 26 April 2024 02:26">Friday 26 April 2024</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
