<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Customizing LD files · GhostKnockoffGWAS</title><meta name="title" content="Customizing LD files · GhostKnockoffGWAS"/><meta property="og:title" content="Customizing LD files · GhostKnockoffGWAS"/><meta property="twitter:title" content="Customizing LD files · GhostKnockoffGWAS"/><meta name="description" content="Documentation for GhostKnockoffGWAS."/><meta property="og:description" content="Documentation for GhostKnockoffGWAS."/><meta property="twitter:description" content="Documentation for GhostKnockoffGWAS."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">GhostKnockoffGWAS</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../intro/">Introduction</a></li><li><a class="tocitem" href="../documentation/">Documentation</a></li><li><a class="tocitem" href="../download/">Downloads</a></li><li><a class="tocitem" href="../zfile/">Acceptable Z-scores file</a></li><li><a class="tocitem" href="../examples/">Tutorial</a></li><li class="is-active"><a class="tocitem" href>Customizing LD files</a><ul class="internal"><li><a class="tocitem" href="#Command-line-documentation-of-solveblock-executable"><span>Command-line documentation of <code>solveblock</code> executable</span></a></li><li><a class="tocitem" href="#Required-inputs"><span>Required inputs</span></a></li><li><a class="tocitem" href="#Optional-inputs"><span>Optional inputs</span></a></li><li><a class="tocitem" href="#Output-format"><span>Output format</span></a></li><li><a class="tocitem" href="#Determining-start_bp-and-end_bp"><span>Determining <code>start_bp</code> and <code>end_bp</code></span></a></li><li><a class="tocitem" href="#A-note-on-run-time"><span>A note on run-time</span></a></li></ul></li><li><a class="tocitem" href="../FAQ/">FAQ</a></li><li><a class="tocitem" href="../julia/">Usage within Julia</a></li><li><a class="tocitem" href="../developer/">Developer documentation</a></li><li><a class="tocitem" href="../gallery/">Gallery</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Customizing LD files</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Customizing LD files</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/GhostKnockoffGWAS/blob/main/docs/src/man/solveblocks.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Customizing-your-own-LD-files"><a class="docs-heading-anchor" href="#Customizing-your-own-LD-files">Customizing your own LD files</a><a id="Customizing-your-own-LD-files-1"></a><a class="docs-heading-anchor-permalink" href="#Customizing-your-own-LD-files" title="Permalink"></a></h1><p>One can customize the <code>--LD-files</code> input starting with individual level data stored in VCF format. This feature is supported by the <code>solveblock</code> executable located within <code>GhostKnockoffGWAS/bin/solveblock</code>.</p><h2 id="Command-line-documentation-of-solveblock-executable"><a class="docs-heading-anchor" href="#Command-line-documentation-of-solveblock-executable">Command-line documentation of <code>solveblock</code> executable</a><a id="Command-line-documentation-of-solveblock-executable-1"></a><a class="docs-heading-anchor-permalink" href="#Command-line-documentation-of-solveblock-executable" title="Permalink"></a></h2><p>Simple run</p><pre><code class="nohighlight hljs">solveblock --vcffile test.vcf.gz --chr 1 --start_bp 10583 --end_bp 1892607 --outdir ./test_LD_files --genome-build 19 </code></pre><h2 id="Required-inputs"><a class="docs-heading-anchor" href="#Required-inputs">Required inputs</a><a id="Required-inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Required-inputs" title="Permalink"></a></h2><table><tr><th style="text-align: left">Option name</th><th style="text-align: center">Argument</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: left"><code>--vcffile</code></td><td style="text-align: center">String</td><td style="text-align: left">A VCF file storing individual level genotypes. Must end in <code>.vcf</code> or <code>.vcf.gz</code>. The ALT field for each record must be unique, i.e. multiallelic records must be split first. Missing genotypes <code>GT</code> will be imputed by column mean.</td></tr><tr><td style="text-align: left"><code>--chr</code></td><td style="text-align: center">Int</td><td style="text-align: left">Target chromosome. This MUST be an integer and it must match the <code>CHROM</code> field in your VCF file. Thus, if your VCF file has CHROM field like <code>chr1</code>, <code>CHR1</code>, or <code>CHROM1</code> etc, each record must be renamed into <code>1</code>.</td></tr><tr><td style="text-align: left"><code>--start_bp</code></td><td style="text-align: center">Int</td><td style="text-align: left">starting basepair (position)</td></tr><tr><td style="text-align: left"><code>--end_bp</code></td><td style="text-align: center">Int</td><td style="text-align: left">ending basepair (position)</td></tr><tr><td style="text-align: left"><code>--outdir</code></td><td style="text-align: center">String</td><td style="text-align: left">Directory that the output will be stored in (must exist)</td></tr><tr><td style="text-align: left"><code>--genome-build</code></td><td style="text-align: center">Int</td><td style="text-align: left">human genome build for position field of the VCF file. This must be 19 (hg19) or 38 (hg38)</td></tr></table><h2 id="Optional-inputs"><a class="docs-heading-anchor" href="#Optional-inputs">Optional inputs</a><a id="Optional-inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Optional-inputs" title="Permalink"></a></h2><table><tr><th style="text-align: left">Option name</th><th style="text-align: center">Argument</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: left"><code>--tol</code></td><td style="text-align: center">Float64</td><td style="text-align: left">Convergence tolerlance for group knockoff coordinate descent optimization (default <code>0.0001</code>)</td></tr><tr><td style="text-align: left"><code>--min_maf</code></td><td style="text-align: center">Float64</td><td style="text-align: left">Minimum minor allele frequency for a variable to be considered (default <code>0.01</code>)</td></tr><tr><td style="text-align: left"><code>--min_hwe</code></td><td style="text-align: center">Int</td><td style="text-align: left">Cutoff for hardy-weinburg equilibrium p-values. Only SNPs with p-value &gt;= <code>min_hwe</code> will be included (default <code>0.0</code>)</td></tr><tr><td style="text-align: left"><code>--method</code></td><td style="text-align: center">String</td><td style="text-align: left">group knockoff optimization algorithm, choices include <code>maxent</code> (defualt), <code>mvr</code>, <code>sdp</code>, or <code>equi</code>. See sec 2 of https://arxiv.org/abs/2310.15069</td></tr><tr><td style="text-align: left"><code>--linkage</code></td><td style="text-align: center">String</td><td style="text-align: left">Linkage function to use for defining group membership. It defines how the distances between features are aggregated into the distances between groups. Valid choices include <code>average</code> (default), <code>single</code>, <code>complete</code>, <code>ward</code>, and <code>ward_presquared</code>. Note if <code>force_contiguous=true</code>, <code>linkage</code> must be <code>:single</code></td></tr><tr><td style="text-align: left"><code>--force_contiguous</code></td><td style="text-align: center">Bool</td><td style="text-align: left">whether to force groups to be contiguous (default <code>false</code>). Note if <code>force_contiguous=true</code>, <code>linkage</code> must be <code>:single</code>)</td></tr><tr><td style="text-align: left"><code>--group_cor_cutoff</code></td><td style="text-align: center">Float64</td><td style="text-align: left">correlation cutoff value for defining groups (default <code>0.5</code>). Value should be between 0 and 1, where larger values correspond to larger groups.</td></tr><tr><td style="text-align: left"><code>--group_rep_cutoff</code></td><td style="text-align: center">Float64</td><td style="text-align: left">cutoff value for selecting group-representatives (default <code>0.5</code>). Value should be between 0 and 1, where larger values correspond to more representatives per group.</td></tr><tr><td style="text-align: left"><code>--verbose</code></td><td style="text-align: center">Bool</td><td style="text-align: left">Whether to print intermediate messages (default <code>true</code>)</td></tr></table><h2 id="Output-format"><a class="docs-heading-anchor" href="#Output-format">Output format</a><a id="Output-format-1"></a><a class="docs-heading-anchor-permalink" href="#Output-format" title="Permalink"></a></h2><p>Calling <code>solveblock</code> will create 3 files in the directory <code>outdir/chr</code> (the <code>chr</code> directory will be created if it doesn&#39;t exist, but <code>outdir</code> must exist):</p><ol><li><code>XXX.h5</code>:  This contains the following data for region XXX<ul><li><code>D</code>: A <code>p × p</code> (dense) matrix corresponding to the S matrix for both the   representative and non-representative variables. Knockoff sampling should    use this matrix. </li><li><code>S</code>: Matrix obtained from solving the group knockoff optimization problem    on the representative (group-key) variables.</li><li><code>Sigma</code>: The original <code>p × p</code> correlation matrix estimated from <code>vcffile</code></li><li><code>SigmaInv</code>: Inverse of <code>Sigma</code></li><li><code>Sigma_reps</code>: The correlation matrix for the representative variables. This   is the matrix actually used to solve the group-knockoff optimization</li><li><code>Sigma_reps_inv</code>: Inverse of <code>Sigma_reps</code></li><li><code>group_reps</code>: Indices <code>groups</code> that are used as representatives (i.e.    group-key variables)</li><li><code>groups</code>: The group membership vector</li></ul></li><li><code>Info_XXX.csv</code>: This includes information for each variant (chr/pos/etc) present   in the corresponding <code>.h5</code> file.</li><li><code>summary_XXX.csv</code>: Summary file for the knockoff optimization problem</li></ol><h2 id="Determining-start_bp-and-end_bp"><a class="docs-heading-anchor" href="#Determining-start_bp-and-end_bp">Determining <code>start_bp</code> and <code>end_bp</code></a><a id="Determining-start_bp-and-end_bp-1"></a><a class="docs-heading-anchor-permalink" href="#Determining-start_bp-and-end_bp" title="Permalink"></a></h2><ul><li>In our papers, we defined each start and end position by adapting the <a href="https://bitbucket.org/nygcresearch/ldetect-data/src/master/">quasi-independent regions of ldetect</a>. </li><li>Given individual level data, one can compute approximately independent LD blocks directly, see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8696101/">reference</a> and <a href="https://privefl.github.io/bigsnpr/reference/snp_ldsplit.html">R software</a>.</li></ul><h2 id="A-note-on-run-time"><a class="docs-heading-anchor" href="#A-note-on-run-time">A note on run-time</a><a id="A-note-on-run-time-1"></a><a class="docs-heading-anchor-permalink" href="#A-note-on-run-time" title="Permalink"></a></h2><p>Because VCF files are plain text files, it is inherently slow to read even if it is indexed. Thus, we <em>strongly recommend</em> one to split the input VCF file by chromosomes, and possibly into smaller chunks, before running <code>solveblock</code>. For optimal performance, it is best to filter the VCF file down to records between <code>start_bp</code> and <code>end_bp</code> (e.g. with <code>bcftools</code>) before running <code>solveblock</code>. </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../examples/">« Tutorial</a><a class="docs-footer-nextpage" href="../FAQ/">FAQ »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.5.0 on <span class="colophon-date" title="Thursday 27 June 2024 02:36">Thursday 27 June 2024</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
